### ISSUE TITLE
tcmalloc deadlock in do_malloc_pages() using SpinLock::SlowLock

### ISSUE THREAD

**Issue by f18m** [gperftools/gperftools#713]
Hi,

I'm using tcmalloc in a rather-large project, running 24hours per day, 7days per week.
It does allocate/deallocate quite a large amount of memory (it runs on an x864_64 machine, running SLES 11.3, kernel is 3.0.76-11 SMP, available RAM is 192GB and this is a NUMA system with dual CPUs).

I found that randomly the software enters a deadlock condition, where one thread is using 100% of CPU doing a tcmalloc of ~1MB:

#0  0x0000000000528e70 in tcmalloc::PageHeap::AllocLarge (this=0x919000, n=137) at src/page_heap.cc:174
#1  0x000000000052970e in tcmalloc::PageHeap::New (this=0x919000, n=137) at src/page_heap.cc:119
#2  0x000000000054ffad in do_malloc_pages (size=1122304, heap=0x9b2620) at src/tcmalloc.cc:1077
#3  do_malloc_no_errno (size=1114160) at src/tcmalloc.cc:1110
#4  do_malloc (size=1114160) at src/tcmalloc.cc:1115
#5  do_malloc_or_cpp_alloc (size=1114160) at src/tcmalloc.cc:1035
#6  tc_malloc (size=1114160) at src/tcmalloc.cc:1574

and several other threads are blocked in do_malloc_pages():

#0  0x000000000052c6a7 in sys_futex (t=0x7ff5457c11d0, v=2, o=128, a=0x8b3fa0) at ./src/base/linux_syscall_support.h:2099
#1  base::internal::SpinLockDelay (w=0x8b3fa0, value=2, loop=<optimized out>) at ./src/base/spinlock_linux-inl.h:88
#2  0x000000000052c54c in SpinLock::SlowLock (this=0x8b3fa0) at src/base/spinlock.cc:133
#3  0x000000000054ff9e in do_malloc_pages (size=1122304, heap=0x9ae220) at src/tcmalloc.cc:1076
#4  do_malloc_no_errno (size=1114160) at src/tcmalloc.cc:1110
#5  do_malloc (size=1114160) at src/tcmalloc.cc:1115
#6  do_malloc_or_cpp_alloc (size=1114160) at src/tcmalloc.cc:1035
#7  tc_malloc (size=1114160) at src/tcmalloc.cc:1574

I'm using gperftools-2.4, and linking libtcmalloc_minimal statically.

Any hint about what's going wrong?

Thanks a lot!

**Comment by alk** [comment]
Given that AllocLarge is not blocked, it is not a deadlock. Likely manifestation of https://github.com/gperftools/gperftools/issues/535

Can you please grab malloc stats from MallocExtension::GetInstance()->GetStats and post it here ?

**Comment by f18m** [comment]
Hi alk,
sorry for the delay with this issue! 

Honestly I am quite sure the problem is not just a slowdown but a real deadlock, since the software where it happens (a server process) stops processing any packet it receives and remains blocked with a thread at 100% forever, until manually killed.
Anyway here is the stats I got after 4 days of uptime of that server:

```
Tcmalloc statistics:
------------------------------------------------
MALLOC:    18135383784 (17295.2 MiB) Bytes in use by application
MALLOC: +    249143296 (  237.6 MiB) Bytes in page heap freelist
MALLOC: +    565623264 (  539.4 MiB) Bytes in central cache freelist
MALLOC: +     20610912 (   19.7 MiB) Bytes in transfer cache freelist
MALLOC: +     23996376 (   22.9 MiB) Bytes in thread cache freelists
MALLOC: +     40038560 (   38.2 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =  19034796192 (18153.0 MiB) Actual memory used (physical + swap)
MALLOC: +            0 (    0.0 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =  19034796192 (18153.0 MiB) Virtual address space used
MALLOC:
MALLOC:         376618              Spans in use
MALLOC:             49              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
```

Now my best guess is that the 539.4 MiB in central cache freelist produce a very high number of locking/unlocking operations of the SpinLock inside do_malloc_pages() function... my guess is that this can generate such  a deadlock situation I am seeing only for 2 reasons (any idea welcome):

1) memory corruption: my server process corrupts somehow the tcmalloc structures and in particular SpinLock internal variables... however this should happen very deterministically and thus appears a bit unlikely.

2) there is a bug in tcmalloc SpinLock implementation... this leads me to the question: why tcmalloc is not using standard pthread spinlocks? is it easy to change tcmalloc sources to use standard pthread spinlocks?

Thanks a lot for any help!!

**Comment by alk** [comment]
Hi. You've posted not all statistics. The should be couple more pages worth of stats. Including per-page-span stats.

You're two ideas cannot be 100% excluded but number 2 is extremely unlikely since other users would have reported it. Homegrown spinlock (which is actually mutex since it sleeps) is used since libpthread mutexes are significantly slower. Homegrown spinlock doesn't have to check various lock types, track owner etc like pthread mutex does. Plus it's fast-path is inlined.

Please post full stats and we should be able to see if my guess is right or not (I think it is).

Also in case it is not what I think, please post backtraces of all threads in case we've missed something else.

**Comment by alk** [comment]
Hm. So looking at stats you've posted it seems that page heap is not large so large page heap freelists should not be a problem.

So please still post statistics (close to time when your program gets stuck). But also try running your program with TCMALLOC_TRANSFER_NUM_OBJ=40.

Also how many threads you have ? If the answer is "a ton" you might also want increase total thread cache size. By setting TCMALLOC_MAX_TOTAL_THREAD_CACHE_BYTES=<2-3 gigabytes>

**Comment by alk** [comment]
I see 49 thread heaps reported so at the time of stats capturing you had "only" 49 threads. Which is not a ton.

But huge count of spans in use indicates that maybe I was right about page heap fragmentation. It could be if you managed to grab stats after massive slowdown.

**Comment by f18m** [comment]
Hi alk,
that's a lot of useful hints, thanks very much.

Just a dumb question: how can I gather more stats, e.g., per-page-span stats? I copied&pasted all the output reported by MallocExtension::GetStats()... 

thanks!

**Comment by alk** [comment]
Just make sure that you're passing buffer that is bigger than 10k. And it'll output more detailed stats (within buffer size limits).

**Comment by alk** [comment]
Also revisiting your original description I see that both threads (and you mention more) are allocating 1M + something chunks. Because you mention that a lot of your threads are stuck in such allocations, it means your program is allocating those 1 meg chunks often.

I should warn you that if that is indeed the case and your program relies on quickly and frequently allocating 1+ meg chunks, then current tcmalloc design won't be fast for your use case. Current tcmalloc is built with assumption that 1+ megs allocations are rare enough to handle them slowly.

Also I'm curious how the program manages to use so big allocations so quickly. Is it the case that those are some kind of vectors that are allocated big to have space for grows? Or hash-tables with large initial sizes (to avoid rehashing)? Is your program actually able to utilize those large allocations all the time (I'm most curious about this one)? If not you might consider starting with smaller allocations. Anything smaller then 1 meg should work much better with current tcmalloc's design.

**Comment by f18m** [comment]
Hi alk,

regarding  1M+ chunks: yes, recently my software was changed and now such allocations are done very often.
Previously a custom memory pool was implemented but since we are using tcmalloc we decided to throw it away
and let tcmalloc handle memory pooling and reuse.
The reason why we do 1M+ allocations so often is that we need such chunks for serialization of in-memory objects
that need to be written on disk. After write, the 1M+ block is released.
Can we tweak some tcmalloc parameter to allow it to consider 1M allocations something that happens often and thus
that needs to be handled quickly, just like smaller allocations?

> post statistics (close to time when your program gets stuck). 
I will post complete stats in the next days, since my feeling is that the stats will converge to "stable values"
only after several hours of server operation...

> But also try running your program with TCMALLOC_TRANSFER_NUM_OBJ=40.
I have restarted it with that setting.

Thanks!

**Comment by alk** [comment]
On Mon, Oct 5, 2015 at 2:33 AM, f18m <notifications@github.com> wrote:

> Hi alk,
>
> regarding 1M+ chunks: yes, recently my software was changed and now such
> allocations are done very often.
> Previously a custom memory pool was implemented but since we are using
> tcmalloc we decided to throw it away
> and let tcmalloc handle memory pooling and reuse.
> The reason why we do 1M+ allocations so often is that we need such chunks
> for serialization of in-memory objects
> that need to be written on disk. After write, the 1M+ block is released.
> Can we tweak some tcmalloc parameter to allow it to consider 1M
> allocations something that happens often and thus
> that needs to be handled quickly, just like smaller allocations?
>

You can try bumping up kMaxPages like that:

diff --git a/src/common.h b/src/common.h
index c3484d3..751f099 100644
--- a/src/common.h
+++ b/src/common.h
@@ -101,7 +101,7 @@ static const size_t kMaxSize    = 256 * 1024;
 static const size_t kAlignment  = 8;
 static const size_t kLargeSizeClass = 0;
 // For all span-lengths < kMaxPages we keep an exact-size list.
-static const size_t kMaxPages = 1 << (20 - kPageShift);
+static const size_t kMaxPages = 1 << (24 - kPageShift);

 // Default bound on the total amount of thread caches.
 #ifdef TCMALLOC_SMALL_BUT_SLOW


But you should really consider doing something with your program design.
Since your current design looks like lots of memory waste. Dynamically
growable buffers for serialization will be fast, amortized O(N) performance
(proportional to size of output) and good on memory waste.

**Comment by f18m** [comment]
Hi alk,

Thanks for your reply. After some discussion internally with my colleagues we decided that, since we know exactly the usage pattern for those 1M+ chunks, we will just malloc() them all at once at the beginning and then re-use them, instead of doing malloc()/free() each time.

I have just 2 more questions:
1) do you think that this allocation of 1M+ chunks is the root origin of page heap fragmentation, leading to a tremendous slow-down (that basically makes a thread stuck at 100% for a long time), that could be the same of #535 ? 
The reason I ask is that this continuous allocation/free of 1M+ chunks is the only thing that changed recently in my software... page heap fragmentation due to other smaller allocations should be at the same level of my older software versions which however did NOT suffer of this deadlock-like situation (and did use tcmalloc too!)... 

2) looking at the kMaxPages setting you pointed out, I noticed the existence of --with-tcmalloc-pagesize option... looking at the stats I previously posted (I will post more stats ASAP), would you reccomend using "--with-tcmalloc-pagesize=32/64" ?

**Comment by alk** [comment]
On Wed, Oct 7, 2015 at 8:34 AM, f18m <notifications@github.com> wrote:

> Hi alk,
>
> Thanks for your reply. After some discussion internally with my colleagues
> we decided that, since we know exactly the usage pattern for those 1M+
> chunks, we will just malloc() them all at once at the beginning and then
> re-use them, instead of doing malloc()/free() each time.
>
> I have just 2 more questions:
> 1) do you think that this allocation of 1M+ chunks is the root origin of
> page heap fragmentation, leading to a tremendous slow-down (that basically
> makes a thread stuck at 100% for a long time), that could be the same of
> #535 <https://github.com/gperftools/gperftools/issues/535> ?
> The reason I ask is that this continuous allocation/free of 1M+ chunks is
> the only thing that changed recently in my software... page heap
> fragmentation due to other smaller allocations should be at the same level
> of my older software versions which however did NOT suffer of this
> deadlock-like situation (and did use tcmalloc too!)...
>

I think so. Looks very plausible based on little data that I have.


> 2) looking at the kMaxPages setting you pointed out, I noticed the
> existence of --with-tcmalloc-pagesize option... looking at the stats I
> previously posted (I will post more stats ASAP), would you reccomend using
> "--with-tcmalloc-pagesize=32/64" ?
>

That should not affect 1+ megs allocation. You should separately evaluate
benefits of larger page size. It could be slightly faster, but it will also
increase memory fragmentation a bit. How much depends on your program and
your hardware.

**Comment by f18m** [comment]
Hi alk,
here are some more stats, on a very loaded server machine (different from the previous one), where 1M+ chunks are recycled and so there are no frequent malloc()/free() calls of size 1M:

```
Tcmalloc statistics:
------------------------------------------------
MALLOC:    37002863824 (35288.7 MiB) Bytes in use by application
MALLOC: +   5207760896 ( 4966.5 MiB) Bytes in page heap freelist
MALLOC: +    657939920 (  627.5 MiB) Bytes in central cache freelist
MALLOC: +      3061344 (    2.9 MiB) Bytes in transfer cache freelist
MALLOC: +     19400448 (   18.5 MiB) Bytes in thread cache freelists
MALLOC: +     93843616 (   89.5 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =  42984870048 (40993.6 MiB) Actual memory used (physical + swap)
MALLOC: +            0 (    0.0 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =  42984870048 (40993.6 MiB) Virtual address space used
MALLOC:
MALLOC:         650175              Spans in use
MALLOC:             56              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
------------------------------------------------
Total size of freelists for per-thread caches,
transfer cache, and central cache, by size class
------------------------------------------------
class   1 [        8 bytes ] :    55548 objs;   0.4 MiB;   0.4 cum MiB
class   2 [       16 bytes ] :     8516 objs;   0.1 MiB;   0.6 cum MiB
class   3 [       32 bytes ] :   817499 objs;  24.9 MiB;  25.5 cum MiB
class   4 [       48 bytes ] :   673910 objs;  30.8 MiB;  56.4 cum MiB
class   5 [       64 bytes ] :  3822577 objs; 233.3 MiB; 289.7 cum MiB
class   6 [       80 bytes ] :   152291 objs;  11.6 MiB; 301.3 cum MiB
class   7 [       96 bytes ] :    70990 objs;   6.5 MiB; 307.8 cum MiB
class   8 [      112 bytes ] :    18867 objs;   2.0 MiB; 309.8 cum MiB
class   9 [      128 bytes ] :     6144 objs;   0.8 MiB; 310.5 cum MiB
class  10 [      144 bytes ] :     9560 objs;   1.3 MiB; 311.9 cum MiB
class  11 [      160 bytes ] :     3829 objs;   0.6 MiB; 312.4 cum MiB
class  12 [      176 bytes ] :    74843 objs;  12.6 MiB; 325.0 cum MiB
class  13 [      192 bytes ] :   426130 objs;  78.0 MiB; 403.0 cum MiB
class  14 [      208 bytes ] :     1880 objs;   0.4 MiB; 403.4 cum MiB
class  15 [      224 bytes ] :     1906 objs;   0.4 MiB; 403.8 cum MiB
class  16 [      240 bytes ] :     1724 objs;   0.4 MiB; 404.2 cum MiB
class  17 [      256 bytes ] :     1437 objs;   0.4 MiB; 404.6 cum MiB
class  18 [      288 bytes ] :     1477 objs;   0.4 MiB; 405.0 cum MiB
class  19 [      320 bytes ] :     1038 objs;   0.3 MiB; 405.3 cum MiB
class  20 [      352 bytes ] :      786 objs;   0.3 MiB; 405.5 cum MiB
class  21 [      384 bytes ] :     1445 objs;   0.5 MiB; 406.1 cum MiB
class  22 [      416 bytes ] :     1604 objs;   0.6 MiB; 406.7 cum MiB
class  23 [      448 bytes ] :     3026 objs;   1.3 MiB; 408.0 cum MiB
class  24 [      480 bytes ] :     2738 objs;   1.3 MiB; 409.3 cum MiB
class  25 [      512 bytes ] :     2554 objs;   1.2 MiB; 410.5 cum MiB
class  26 [      576 bytes ] :     2458 objs;   1.4 MiB; 411.9 cum MiB
class  27 [      640 bytes ] :     2547 objs;   1.6 MiB; 413.4 cum MiB
class  28 [      704 bytes ] :     7292 objs;   4.9 MiB; 418.3 cum MiB
class  29 [      768 bytes ] :     2297 objs;   1.7 MiB; 420.0 cum MiB
class  30 [      832 bytes ] :     1420 objs;   1.1 MiB; 421.1 cum MiB
class  31 [      896 bytes ] :     1100 objs;   0.9 MiB; 422.1 cum MiB
class  32 [      960 bytes ] :    84685 objs;  77.5 MiB; 499.6 cum MiB
class  33 [     1024 bytes ] :      859 objs;   0.8 MiB; 500.4 cum MiB
class  34 [     1152 bytes ] :      822 objs;   0.9 MiB; 501.3 cum MiB
class  35 [     1280 bytes ] :      705 objs;   0.9 MiB; 502.2 cum MiB
class  36 [     1408 bytes ] :     2135 objs;   2.9 MiB; 505.1 cum MiB
class  37 [     1536 bytes ] :      410 objs;   0.6 MiB; 505.7 cum MiB
class  38 [     1792 bytes ] :      526 objs;   0.9 MiB; 506.6 cum MiB
class  39 [     2048 bytes ] :       52 objs;   0.1 MiB; 506.7 cum MiB
class  40 [     2304 bytes ] :        8 objs;   0.0 MiB; 506.7 cum MiB
class  41 [     2560 bytes ] :       11 objs;   0.0 MiB; 506.7 cum MiB
class  42 [     2816 bytes ] :        6 objs;   0.0 MiB; 506.7 cum MiB
class  43 [     3072 bytes ] :        9 objs;   0.0 MiB; 506.7 cum MiB
class  44 [     3328 bytes ] :        5 objs;   0.0 MiB; 506.8 cum MiB
class  45 [     4096 bytes ] :       11 objs;   0.0 MiB; 506.8 cum MiB
class  46 [     4608 bytes ] :       13 objs;   0.1 MiB; 506.9 cum MiB
class  47 [     5120 bytes ] :        6 objs;   0.0 MiB; 506.9 cum MiB
class  48 [     6144 bytes ] :        8 objs;   0.0 MiB; 506.9 cum MiB
class  49 [     6656 bytes ] :       10 objs;   0.1 MiB; 507.0 cum MiB
class  50 [     8192 bytes ] :       13 objs;   0.1 MiB; 507.1 cum MiB
class  51 [     9216 bytes ] :        6 objs;   0.1 MiB; 507.2 cum MiB
class  52 [    10240 bytes ] :        4 objs;   0.0 MiB; 507.2 cum MiB
class  53 [    12288 bytes ] :        8 objs;   0.1 MiB; 507.3 cum MiB
class  54 [    13312 bytes ] :        3 objs;   0.0 MiB; 507.3 cum MiB
class  55 [    16384 bytes ] :       17 objs;   0.3 MiB; 507.6 cum MiB
class  56 [    20480 bytes ] :        2 objs;   0.0 MiB; 507.6 cum MiB
class  57 [    24576 bytes ] :        3 objs;   0.1 MiB; 507.7 cum MiB
class  59 [    32768 bytes ] :       17 objs;   0.5 MiB; 508.2 cum MiB
class  60 [    40960 bytes ] :        1 objs;   0.0 MiB; 508.3 cum MiB
class  62 [    57344 bytes ] :        1 objs;   0.1 MiB; 508.3 cum MiB
class  76 [   172032 bytes ] :        1 objs;   0.2 MiB; 508.5 cum MiB
class  87 [   262144 bytes ] :        1 objs;   0.2 MiB; 508.7 cum MiB
------------------------------------------------
PageHeap: 93 sizes; 4966.5 MiB free;    0.0 MiB unmapped
------------------------------------------------
     1 pages *    351 spans ~    2.7 MiB;    2.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     2 pages *   3479 spans ~   54.4 MiB;   57.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     3 pages *     26 spans ~    0.6 MiB;   57.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     4 pages *   3315 spans ~  103.6 MiB;  161.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     5 pages *      7 spans ~    0.3 MiB;  161.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     6 pages *   3995 spans ~  187.3 MiB;  348.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     7 pages *      2 spans ~    0.1 MiB;  349.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     8 pages *   2718 spans ~  169.9 MiB;  518.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     9 pages *      2 spans ~    0.1 MiB;  519.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    10 pages *   2062 spans ~  161.1 MiB;  680.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    11 pages *      1 spans ~    0.1 MiB;  680.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    12 pages *   1743 spans ~  163.4 MiB;  843.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    13 pages *      1 spans ~    0.1 MiB;  843.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    14 pages *   1529 spans ~  167.2 MiB; 1010.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    15 pages *      3 spans ~    0.4 MiB; 1011.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    16 pages *   1212 spans ~  151.5 MiB; 1162.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    17 pages *      1 spans ~    0.1 MiB; 1162.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    18 pages *    974 spans ~  137.0 MiB; 1299.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    19 pages *      2 spans ~    0.3 MiB; 1300.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    20 pages *    883 spans ~  138.0 MiB; 1438.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    21 pages *      1 spans ~    0.2 MiB; 1438.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    22 pages *   1070 spans ~  183.9 MiB; 1622.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    23 pages *      1 spans ~    0.2 MiB; 1622.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    24 pages *    581 spans ~  108.9 MiB; 1731.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    25 pages *      1 spans ~    0.2 MiB; 1731.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    26 pages *    486 spans ~   98.7 MiB; 1830.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    27 pages *      1 spans ~    0.2 MiB; 1830.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    28 pages *    422 spans ~   92.3 MiB; 1922.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    30 pages *    418 spans ~   98.0 MiB; 2020.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    32 pages *    356 spans ~   89.0 MiB; 2109.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    33 pages *      2 spans ~    0.5 MiB; 2110.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    34 pages *    356 spans ~   94.6 MiB; 2204.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    35 pages *      1 spans ~    0.3 MiB; 2205.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    36 pages *    285 spans ~   80.2 MiB; 2285.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    37 pages *      1 spans ~    0.3 MiB; 2285.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    38 pages *    332 spans ~   98.6 MiB; 2384.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    40 pages *    231 spans ~   72.2 MiB; 2456.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    42 pages *    194 spans ~   63.7 MiB; 2519.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    44 pages *    203 spans ~   69.8 MiB; 2589.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    45 pages *      1 spans ~    0.4 MiB; 2590.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    46 pages *    178 spans ~   64.0 MiB; 2654.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    48 pages *    167 spans ~   62.6 MiB; 2716.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    50 pages *    181 spans ~   70.7 MiB; 2787.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    52 pages *    150 spans ~   60.9 MiB; 2848.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    53 pages *      2 spans ~    0.8 MiB; 2849.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    54 pages *    126 spans ~   53.2 MiB; 2902.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    55 pages *      1 spans ~    0.4 MiB; 2902.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    56 pages *    124 spans ~   54.2 MiB; 2956.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    58 pages *    139 spans ~   63.0 MiB; 3019.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    59 pages *      1 spans ~    0.5 MiB; 3020.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    60 pages *    113 spans ~   53.0 MiB; 3073.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    61 pages *      1 spans ~    0.5 MiB; 3073.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    62 pages *    124 spans ~   60.1 MiB; 3133.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    64 pages *     96 spans ~   48.0 MiB; 3181.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    66 pages *     82 spans ~   42.3 MiB; 3224.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    68 pages *     91 spans ~   48.3 MiB; 3272.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    70 pages *    117 spans ~   64.0 MiB; 3336.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    72 pages *    101 spans ~   56.8 MiB; 3393.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    74 pages *     71 spans ~   41.0 MiB; 3434.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    76 pages *     89 spans ~   52.8 MiB; 3487.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    77 pages *      1 spans ~    0.6 MiB; 3487.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    78 pages *     85 spans ~   51.8 MiB; 3539.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    79 pages *      1 spans ~    0.6 MiB; 3540.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    80 pages *     97 spans ~   60.6 MiB; 3600.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    82 pages *     84 spans ~   53.8 MiB; 3654.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    83 pages *      1 spans ~    0.6 MiB; 3655.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    84 pages *     87 spans ~   57.1 MiB; 3712.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    86 pages *     86 spans ~   57.8 MiB; 3770.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    87 pages *      1 spans ~    0.7 MiB; 3770.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    88 pages *     89 spans ~   61.2 MiB; 3832.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    90 pages *     82 spans ~   57.7 MiB; 3889.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    92 pages *     86 spans ~   61.8 MiB; 3951.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    94 pages *     80 spans ~   58.8 MiB; 4010.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    95 pages *      1 spans ~    0.7 MiB; 4011.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    96 pages *     73 spans ~   54.8 MiB; 4065.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    97 pages *      1 spans ~    0.8 MiB; 4066.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    98 pages *     52 spans ~   39.8 MiB; 4106.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   100 pages *     44 spans ~   34.4 MiB; 4140.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   102 pages *     41 spans ~   32.7 MiB; 4173.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   103 pages *      1 spans ~    0.8 MiB; 4174.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   104 pages *     40 spans ~   32.5 MiB; 4206.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   106 pages *     30 spans ~   24.8 MiB; 4231.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   108 pages *     41 spans ~   34.6 MiB; 4266.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   110 pages *     30 spans ~   25.8 MiB; 4291.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   112 pages *     30 spans ~   26.2 MiB; 4318.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   114 pages *     28 spans ~   24.9 MiB; 4343.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   116 pages *     27 spans ~   24.5 MiB; 4367.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   118 pages *     33 spans ~   30.4 MiB; 4398.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   120 pages *     24 spans ~   22.5 MiB; 4420.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   121 pages *      1 spans ~    0.9 MiB; 4421.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   122 pages *     32 spans ~   30.5 MiB; 4451.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   124 pages *     15 spans ~   14.5 MiB; 4466.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   126 pages *     21 spans ~   20.7 MiB; 4487.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
>255   large *    342 spans ~  479.4 MiB; 4966.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
```

PS: These stats are after 17hours of operation.

**Comment by alk** [comment]
On Fri, Oct 9, 2015 at 2:43 AM, f18m <notifications@github.com> wrote:

> Hi alk,
> here are some more stats, on a very loaded server machine (different from
> the previous one), where 1M+ chunks are recycled and so there are no
> frequent malloc()/free() calls of size 1M:
>

Hi.

This doesn't look bad. 300+ spans in large spans list is a bit on the
larger side especially if you have more threads and cores, but not as bad
as #535.

Right now I have different things to address on top of my list, but I still
expect to deal with large spans in not too distant future.

Tcmalloc statistics:

**Comment by f18m** [comment]
Hi alk,

> This doesn't look bad. 300+ spans in large spans list is a bit on the
larger side especially if you have more threads and cores, but not as bad
as #535.

I read the comments for #535 but I couldn't find a reference to the number of spans that triggers the performance issue... can you provide me an estimate or just the magnitude order? is it 1k spans or 100k spans that slow downs the tcmalloc to a deadlock-like situation?

Btw the situation on a machine that is doing frequent 1M+ chunk allocations/deallocations, after 2days and 12h of uptime is:

```
Tcmalloc statistics:
------------------------------------------------
MALLOC:    32002792976 (30520.2 MiB) Bytes in use by application
MALLOC: +   1872551936 ( 1785.8 MiB) Bytes in page heap freelist
MALLOC: +    826325688 (  788.0 MiB) Bytes in central cache freelist
MALLOC: +     15326624 (   14.6 MiB) Bytes in transfer cache freelist
MALLOC: +     23464344 (   22.4 MiB) Bytes in thread cache freelists
MALLOC: +     68546720 (   65.4 MiB) Bytes in malloc metadata
MALLOC:   ------------
MALLOC: =  34809008288 (33196.5 MiB) Actual memory used (physical + swap)
MALLOC: +            0 (    0.0 MiB) Bytes released to OS (aka unmapped)
MALLOC:   ------------
MALLOC: =  34809008288 (33196.5 MiB) Virtual address space used
MALLOC:
MALLOC:         561153              Spans in use
MALLOC:             60              Thread heaps in use
MALLOC:           8192              Tcmalloc page size
------------------------------------------------
Call ReleaseFreeMemory() to release freelist memory to the OS (via madvise()).
Bytes released to the OS take up virtual address space but no physical memory.
------------------------------------------------
Total size of freelists for per-thread caches,
transfer cache, and central cache, by size class
------------------------------------------------
class   1 [        8 bytes ] :    79926 objs;   0.6 MiB;   0.6 cum MiB
class   2 [       16 bytes ] :     8425 objs;   0.1 MiB;   0.7 cum MiB
class   3 [       32 bytes ] :   826036 objs;  25.2 MiB;  25.9 cum MiB
class   4 [       48 bytes ] :  2438225 objs; 111.6 MiB; 137.6 cum MiB
class   5 [       64 bytes ] :  1981075 objs; 120.9 MiB; 258.5 cum MiB
class   6 [       80 bytes ] :   405775 objs;  31.0 MiB; 289.4 cum MiB
class   7 [       96 bytes ] :   116250 objs;  10.6 MiB; 300.1 cum MiB
class   8 [      112 bytes ] :    34705 objs;   3.7 MiB; 303.8 cum MiB
class   9 [      128 bytes ] :    36741 objs;   4.5 MiB; 308.3 cum MiB
class  10 [      144 bytes ] :    14971 objs;   2.1 MiB; 310.3 cum MiB
class  11 [      160 bytes ] :     4518 objs;   0.7 MiB; 311.0 cum MiB
class  12 [      176 bytes ] :   139505 objs;  23.4 MiB; 334.4 cum MiB
class  13 [      192 bytes ] :    41717 objs;   7.6 MiB; 342.1 cum MiB
class  14 [      208 bytes ] :     1720 objs;   0.3 MiB; 342.4 cum MiB
class  15 [      224 bytes ] :     1561 objs;   0.3 MiB; 342.7 cum MiB
class  16 [      240 bytes ] :     1113 objs;   0.3 MiB; 343.0 cum MiB
class  17 [      256 bytes ] :     1251 objs;   0.3 MiB; 343.3 cum MiB
class  18 [      288 bytes ] :     1034 objs;   0.3 MiB; 343.6 cum MiB
class  19 [      320 bytes ] :      728 objs;   0.2 MiB; 343.8 cum MiB
class  20 [      352 bytes ] :     2994 objs;   1.0 MiB; 344.8 cum MiB
class  21 [      384 bytes ] :     5828 objs;   2.1 MiB; 346.9 cum MiB
class  22 [      416 bytes ] :     6462 objs;   2.6 MiB; 349.5 cum MiB
class  23 [      448 bytes ] :     2226 objs;   1.0 MiB; 350.5 cum MiB
class  24 [      480 bytes ] :     2336 objs;   1.1 MiB; 351.5 cum MiB
class  25 [      512 bytes ] :     2101 objs;   1.0 MiB; 352.6 cum MiB
class  26 [      576 bytes ] :     5334 objs;   2.9 MiB; 355.5 cum MiB
class  27 [      640 bytes ] :     2010 objs;   1.2 MiB; 356.7 cum MiB
class  28 [      704 bytes ] :    20320 objs;  13.6 MiB; 370.4 cum MiB
class  29 [      768 bytes ] :     1567 objs;   1.1 MiB; 371.5 cum MiB
class  30 [      832 bytes ] :     1056 objs;   0.8 MiB; 372.3 cum MiB
class  31 [      896 bytes ] :      916 objs;   0.8 MiB; 373.1 cum MiB
class  32 [      960 bytes ] :    89514 objs;  82.0 MiB; 455.1 cum MiB
class  33 [     1024 bytes ] :      729 objs;   0.7 MiB; 455.8 cum MiB
class  34 [     1152 bytes ] :    59050 objs;  64.9 MiB; 520.7 cum MiB
class  35 [     1280 bytes ] :   118900 objs; 145.1 MiB; 665.8 cum MiB
class  36 [     1408 bytes ] :      910 objs;   1.2 MiB; 667.0 cum MiB
class  37 [     1536 bytes ] :      468 objs;   0.7 MiB; 667.7 cum MiB
class  38 [     1792 bytes ] :     1118 objs;   1.9 MiB; 669.6 cum MiB
class  39 [     2048 bytes ] :       63 objs;   0.1 MiB; 669.7 cum MiB
class  40 [     2304 bytes ] :       22 objs;   0.0 MiB; 669.8 cum MiB
class  41 [     2560 bytes ] :       28 objs;   0.1 MiB; 669.9 cum MiB
class  42 [     2816 bytes ] :       14 objs;   0.0 MiB; 669.9 cum MiB
class  43 [     3072 bytes ] :       10 objs;   0.0 MiB; 669.9 cum MiB
class  44 [     3328 bytes ] :        4 objs;   0.0 MiB; 669.9 cum MiB
class  45 [     4096 bytes ] :       11 objs;   0.0 MiB; 670.0 cum MiB
class  46 [     4608 bytes ] :       13 objs;   0.1 MiB; 670.0 cum MiB
class  47 [     5120 bytes ] :        3 objs;   0.0 MiB; 670.1 cum MiB
class  48 [     6144 bytes ] :       16 objs;   0.1 MiB; 670.2 cum MiB
class  49 [     6656 bytes ] :       16 objs;   0.1 MiB; 670.3 cum MiB
class  50 [     8192 bytes ] :       22 objs;   0.2 MiB; 670.4 cum MiB
class  51 [     9216 bytes ] :       10 objs;   0.1 MiB; 670.5 cum MiB
class  52 [    10240 bytes ] :       10 objs;   0.1 MiB; 670.6 cum MiB
class  53 [    12288 bytes ] :        6 objs;   0.1 MiB; 670.7 cum MiB
class  54 [    13312 bytes ] :        6 objs;   0.1 MiB; 670.8 cum MiB
class  55 [    16384 bytes ] :       13 objs;   0.2 MiB; 671.0 cum MiB
class  56 [    20480 bytes ] :        4 objs;   0.1 MiB; 671.0 cum MiB
class  57 [    24576 bytes ] :        1 objs;   0.0 MiB; 671.1 cum MiB
class  58 [    26624 bytes ] :        4 objs;   0.1 MiB; 671.2 cum MiB
class  59 [    32768 bytes ] :       33 objs;   1.0 MiB; 672.2 cum MiB
class  60 [    40960 bytes ] :        1 objs;   0.0 MiB; 672.2 cum MiB
class  62 [    57344 bytes ] :        1 objs;   0.1 MiB; 672.3 cum MiB
class  77 [   180224 bytes ] :        1 objs;   0.2 MiB; 672.5 cum MiB
class  87 [   262144 bytes ] :        1 objs;   0.2 MiB; 672.7 cum MiB
------------------------------------------------
PageHeap: 105 sizes; 1785.8 MiB free;    0.0 MiB unmapped
------------------------------------------------
     1 pages *   3744 spans ~   29.2 MiB;   29.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     2 pages *  17392 spans ~  271.8 MiB;  301.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     3 pages *    353 spans ~    8.3 MiB;  309.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     4 pages *   1883 spans ~   58.8 MiB;  368.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     5 pages *     50 spans ~    2.0 MiB;  370.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     6 pages *    542 spans ~   25.4 MiB;  395.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     7 pages *     67 spans ~    3.7 MiB;  399.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     8 pages *   1078 spans ~   67.4 MiB;  466.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
     9 pages *     45 spans ~    3.2 MiB;  469.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    10 pages *    755 spans ~   59.0 MiB;  528.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    11 pages *     18 spans ~    1.5 MiB;  530.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    12 pages *    453 spans ~   42.5 MiB;  572.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    13 pages *     10 spans ~    1.0 MiB;  573.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    14 pages *    332 spans ~   36.3 MiB;  610.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    15 pages *     12 spans ~    1.4 MiB;  611.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    16 pages *    295 spans ~   36.9 MiB;  648.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    17 pages *     11 spans ~    1.5 MiB;  649.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    18 pages *    366 spans ~   51.5 MiB;  701.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    19 pages *      6 spans ~    0.9 MiB;  702.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    20 pages *    291 spans ~   45.5 MiB;  747.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    21 pages *      7 spans ~    1.1 MiB;  748.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    22 pages *    580 spans ~   99.7 MiB;  848.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    23 pages *      6 spans ~    1.1 MiB;  849.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    24 pages *    193 spans ~   36.2 MiB;  885.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    25 pages *      6 spans ~    1.2 MiB;  886.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    26 pages *     76 spans ~   15.4 MiB;  902.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    27 pages *      2 spans ~    0.4 MiB;  902.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    28 pages *     59 spans ~   12.9 MiB;  915.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    29 pages *      4 spans ~    0.9 MiB;  916.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    30 pages *     49 spans ~   11.5 MiB;  928.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    31 pages *      3 spans ~    0.7 MiB;  928.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    32 pages *     39 spans ~    9.8 MiB;  938.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    33 pages *      3 spans ~    0.8 MiB;  939.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    34 pages *     46 spans ~   12.2 MiB;  951.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    35 pages *      2 spans ~    0.5 MiB;  952.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    36 pages *     36 spans ~   10.1 MiB;  962.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    37 pages *      5 spans ~    1.4 MiB;  963.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    38 pages *     39 spans ~   11.6 MiB;  975.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    39 pages *      3 spans ~    0.9 MiB;  976.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    40 pages *     28 spans ~    8.8 MiB;  984.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    41 pages *      4 spans ~    1.3 MiB;  986.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    42 pages *     26 spans ~    8.5 MiB;  994.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    43 pages *      3 spans ~    1.0 MiB;  995.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    44 pages *     28 spans ~    9.6 MiB; 1005.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    45 pages *      3 spans ~    1.1 MiB; 1006.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    46 pages *     33 spans ~   11.9 MiB; 1018.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    47 pages *      2 spans ~    0.7 MiB; 1018.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    48 pages *     30 spans ~   11.2 MiB; 1030.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    49 pages *      5 spans ~    1.9 MiB; 1032.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    50 pages *     13 spans ~    5.1 MiB; 1037.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    51 pages *      4 spans ~    1.6 MiB; 1038.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    52 pages *      8 spans ~    3.2 MiB; 1042.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    54 pages *     27 spans ~   11.4 MiB; 1053.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    55 pages *      1 spans ~    0.4 MiB; 1053.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    56 pages *     19 spans ~    8.3 MiB; 1062.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    57 pages *      6 spans ~    2.7 MiB; 1064.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    58 pages *      9 spans ~    4.1 MiB; 1068.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    59 pages *      4 spans ~    1.8 MiB; 1070.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    60 pages *     16 spans ~    7.5 MiB; 1078.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    61 pages *      2 spans ~    1.0 MiB; 1079.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    62 pages *     13 spans ~    6.3 MiB; 1085.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    63 pages *      1 spans ~    0.5 MiB; 1086.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    64 pages *      8 spans ~    4.0 MiB; 1090.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    65 pages *      1 spans ~    0.5 MiB; 1090.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    66 pages *      6 spans ~    3.1 MiB; 1093.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    68 pages *      6 spans ~    3.2 MiB; 1096.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    69 pages *      1 spans ~    0.5 MiB; 1097.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    70 pages *      5 spans ~    2.7 MiB; 1100.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    71 pages *      1 spans ~    0.6 MiB; 1100.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    72 pages *      9 spans ~    5.1 MiB; 1105.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    74 pages *      7 spans ~    4.0 MiB; 1109.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    75 pages *      2 spans ~    1.2 MiB; 1110.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    76 pages *      6 spans ~    3.6 MiB; 1114.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    78 pages *      7 spans ~    4.3 MiB; 1118.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    79 pages *      1 spans ~    0.6 MiB; 1119.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    80 pages *      2 spans ~    1.2 MiB; 1120.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    82 pages *      5 spans ~    3.2 MiB; 1123.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    84 pages *      2 spans ~    1.3 MiB; 1125.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    85 pages *      1 spans ~    0.7 MiB; 1125.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    86 pages *      2 spans ~    1.3 MiB; 1127.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    87 pages *      2 spans ~    1.4 MiB; 1128.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    88 pages *      5 spans ~    3.4 MiB; 1131.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    90 pages *      1 spans ~    0.7 MiB; 1132.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    92 pages *      1 spans ~    0.7 MiB; 1133.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    94 pages *      3 spans ~    2.2 MiB; 1135.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    95 pages *      1 spans ~    0.7 MiB; 1136.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    97 pages *      2 spans ~    1.5 MiB; 1137.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
    98 pages *      5 spans ~    3.8 MiB; 1141.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   100 pages *      1 spans ~    0.8 MiB; 1142.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   102 pages *      1 spans ~    0.8 MiB; 1143.2 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   104 pages *     10 spans ~    8.1 MiB; 1151.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   106 pages *      3 spans ~    2.5 MiB; 1153.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   107 pages *      1 spans ~    0.8 MiB; 1154.6 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   108 pages *      1 spans ~    0.8 MiB; 1155.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   110 pages *      3 spans ~    2.6 MiB; 1158.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   111 pages *      1 spans ~    0.9 MiB; 1158.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   112 pages *      2 spans ~    1.8 MiB; 1160.7 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   114 pages *      2 spans ~    1.8 MiB; 1162.5 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   115 pages *      1 spans ~    0.9 MiB; 1163.4 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   118 pages *      1 spans ~    0.9 MiB; 1164.3 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   120 pages *      2 spans ~    1.9 MiB; 1166.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   121 pages *      1 spans ~    0.9 MiB; 1167.1 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   122 pages *      2 spans ~    1.9 MiB; 1169.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   123 pages *      1 spans ~    1.0 MiB; 1170.0 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
   126 pages *      2 spans ~    2.0 MiB; 1171.9 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
>255   large *    243 spans ~  613.9 MiB; 1785.8 MiB cum; unmapped:    0.0 MiB;    0.0 MiB cum
```

Here there are "just" 243 spans in the >255 list but I see as alarming the fact that there is 1785.8 MiB in page heap freelist... on the systems modified to avoid frequent 1M+ chunk allocations/deallocations that number is around 500MiB...

**Comment by alk** [comment]
1.7 gigs in page heap freelist looks like for 32 gig total process size.

We still don't have enough evidence about cause of frequent locking. Can you please gather cpu profile from slow system and post it here.

Also I'm not sure I fully understand whether stats posted above are from system that is in this slow state. Looks like number of large spans is relatively low so probably not. Can you please confirm?

I'd like to see stats dump from system that is in slow state.

Regarding now many spans makes it slow, I can't say for sure. I'd say hundreds should be still fine. But thousands are suspicious.

**Comment by alk** [comment]
Since this is likely duplicate of O(N) problem in page heap, I'm closing this ticket.
